client/homebrew/utils/versionHistory.js

```diff
@@ -0,0 +1,120 @@
+export const HISTORY_PREFIX = 'HOMEBREWERY-HISTORY'
+// const HISTORY_SAVE_DELAYS = {
+//  0: 0,           // 0 minutes (if not specified)
+// 	1: 2,			// 2 minutes
+// 	2: 10,			// 10 minutes
+// 	3: 60,			// 60 minutes
+// 	4: 12 * 60,     // 12 hours
+// 	5: 2 * 24 * 60  // 2 days
+// };
+// 
+// const GARBAGE_COLLECT_DELAY = 28 * 24 * 60; // 28 days
+
+
+// <=================== TEST VALUES STARTS ===================>
+
+// Test values
+const HISTORY_SAVE_DELAYS = {
+    0: 0,           // 0 minutes (if not specified)
+	1: 1,			// 1 minutes
+	2: 2,			// 2 minutes
+	3: 3,			// 3 minutes
+	4: 4,           // 4 minutes
+	5: 5            // 5 minutes
+};
+const GARBAGE_COLLECT_DELAY = 10; // 10 minutes
+
+// <==================== TEST VALUES ENDS ====================>
+
+
+
+const DEFAULT_STORED_BREW = {
+    shareId  : 'default_brew',
+    expireAt : '2000-01-01T00:00:00.000Z'
+};
+
+function getKeyBySlot(brew, slot){
+    return `${HISTORY_PREFIX}-${brew.shareId}-${slot}`;
+};
+
+function getVersionBySlot(brew, slot){
+    // Read stored brew data
+    // - If it exists, parse data to object
+    // - If it doesn't exist, pass default object
+    const key = getKeyBySlot(brew, slot);
+    const storedVersion = localStorage.getItem(key);
+    const output = storedVersion ? JSON.parse(storedVersion) : { ...DEFAULT_STORED_BREW, ...brew };
+    return output;
+};
+
+function updateStoredBrew(brew, slot=0){
+    const archiveBrew = {};
+
+    // Data from brew to be stored
+    archiveBrew.title = brew.title;
+    archiveBrew.text = brew.text;
+    archiveBrew.style = brew.style;
+    archiveBrew.version = brew.version;
+    archiveBrew.shareId = brew.shareId;
+
+    // Calculated values
+    archiveBrew.savedAt = new Date();
+    archiveBrew.expireAt = new Date();
+    archiveBrew.expireAt.setMinutes(archiveBrew.expireAt.getMinutes() + HISTORY_SAVE_DELAYS[slot]);
+    
+    // Store data
+    const key = getKeyBySlot(brew, slot);
+    localStorage.setItem(key, JSON.stringify(archiveBrew));
+    return;
+};
+
+export function updateHistory(brew) {
```

Comment:
I think ultimately the logic is correct here, but there's got to be a more straightforward way to do what is effectively just iterating over your `history` array and moving entries around.

Rather than carrying around this `numbers` array, I think things will become a lot simpler if you just use normal for loops and have a `maxSlot` number.

Reply:
Not tested, but I think this might be equivalent:

```
export function updateHistory(brew) {
    const maxSlot = 5;
    const history = [];

    // Load data from local storage to History object
    for (let i = 1; i <= maxSlot; i++) {
        history[i] = getVersionBySlot(brew, i);
    }

    // Walk each version oldest to newest
    for (let slot = maxSlot; slot >= 1; slot--) {

        // If slot has expired, update all lower slots and break
        if (new Date() >= new Date(history[slot].expireAt)) {
            for (let i = slot - 1; i >= 1; i--) {
                updateStoredBrew(history[i], i + 1);	// Move data from slot i to slot i + 1
            }
            updateStoredBrew(brew, 1);	// Fill the first slot with current data
            break;					// Break out of the loop as we found an expired value
        }
    }
}
```